### ISSUE TITLE
[BUG] Wrong attention gradient when Context Parallelism is on

### ISSUE THREAD

**Issue by haitian-jiang** [NVIDIA/Megatron-LM#1464]
**Describe the bug**

- bug location

  When training with context parallelism, the gradient of `core_attention` is wrong.  In another word, the dQ, dK, and dV ourput from the backward function is wrong.  To be more specific, this attention backward function outputs wrong results: `TransformerEngine/transformer_engine/pytorch/attention.py: AttnFuncWithCPAndKVP2P.backward()`.

- how to trigger

  When context parallelism is on and uses this P2P attention module, this bug will be triggered.

- bug behavior

  Given the exactly same parameter and data input for a single card training and a training with context parallelism, the activation gradient of core attention (dQ, dK, dV) do not match in value.  The relative error of these three tensor can be as large as 1.2.  The single card version is double checked with tesnor parallelism and is correct.  

  This will cause the following activation gradient to be wrong in the `linear_qkv` module, and its activation gradient will be used to calculate its parameter gradient.  This error will also accumulate and pollute the gradient calculation of following layers.  The wrong parameter gradient will then accumulate with different micro batches, thus making the `main_grad` for weight update to be wrong, and their relative error will be as large as 2.3.

  So this bug will make the training with context parallelism totally wrong.

**To Reproduce**

Train the GPT model with context parallelism.

1. According to https://github.com/NVIDIA/Megatron-LM/issues/1151, set environment variable `NVTE_BATCH_MHA_P2P_COMM` to 1 to enable P2P communication.

2. Then run `pretrain_gpt.py` with `--context-parallel-size` > 1.

3. Detailed training command:

   ```bash
   CUDA_DEVICE_MAX_CONNECTIONS=1 \
   OMP_NUM_THREADS=1 \
   NVTE_BATCH_MHA_P2P_COMM=1 \
   torchrun \
   --nproc_per_node 2 \
   --nnodes 1 \
   --node_rank 0 \
   --master_addr localhost \
   --master_port 6000 pretrain_gpt.py \
   --tensor-model-parallel-size 1 \
   --pipeline-model-parallel-size 1 \
   --context-parallel-size 2 \  # as long as cp_size > 1
   --num-layers 24 \
   --hidden-size 1024 \
   --num-attention-heads 16 \
   --seq-length 1024 \
   --max-position-embeddings 1024 \
   --micro-batch-size 4 \
   --global-batch-size 16 \
   --lr 0.00015 \
   --train-iters 1 \
   --lr-decay-iters 320000 \
   --lr-decay-style cosine \
   --min-lr 1.0e-5 \
   --weight-decay 1e-2 \
   --lr-warmup-fraction .01 \
   --clip-grad 1.0 \
   --use-mcore-models \
   --attention-softmax-in-fp32 \
   --attention-dropout 0.0 \
   --hidden-dropout 0.0 \
   --bf16 \
   --data-path /workspace/dataset/wikitext/wikitext_text_document \
   --vocab-file /workspace/dataset/gpt2-vocab.json \
   --merge-file /workspace/dataset/gpt2-merges.txt \
   --split 949,50,1 \
   --log-interval 100 \
   --save-interval 10000 \
   --eval-interval 1000 \
   --eval-iters 0 \
   --save /workspace/checkpoints \
   --load /workspace/checkpoints \
   --no-gradient-accumulation-fusion \
   --transformer-impl transformer_engine
   ```

**Expected behavior**

The gradient of attention should be the same as single card when context parallelism is on. 

**Stack trace/logs**

If applicable, add the stack trace or logs from the time of the error.

This is not stack trace or log from console, but the probe of the gradient tensors.

- activation gradient: 

```
name                                                       relative_error  
layers.23|g_activ.output.0                                3.24e-3     0.4ε
layers.23.mlp.linear_fc2|g_activ.output.0                 3.24e-3     0.4ε
layers.23.mlp.linear_fc2|g_activ.input.0                  3.93e-3     0.5ε
layers.23.mlp.linear_fc1|g_activ.output.0                 9.12e-3     1.2ε
layers.23.self_attention.linear_proj|g_activ.output.0     3.90e-3     0.5ε
layers.23.self_attention.core_attention|g_activ.output.0  4.52e-3     0.6ε
layers.23.self_attention.core_attention|g_activ.input.0   1.14e+0   147.3ε
layers.23.self_attention.core_attention|g_activ.input.1   6.41e-1    82.3ε
layers.23.self_attention.core_attention|g_activ.input.2   6.97e-1    89.4ε
layers.23.self_attention.linear_qkv|g_activ.output.0      7.08e-1    90.8ε
layers.23.input_layernorm|g_activ.input.0                 6.78e-1    87.0ε
layers.22|g_activ.output.0                                2.26e-2     2.9ε
layers.22.mlp.linear_fc2|g_activ.output.0                 2.26e-2     2.9ε
layers.22.mlp.linear_fc2|g_activ.input.0                  2.28e-2     2.9ε
layers.22.mlp.linear_fc1|g_activ.output.0                 2.44e-2     3.1ε
layers.22.self_attention.linear_proj|g_activ.output.0     2.27e-2     2.9ε
layers.22.self_attention.core_attention|g_activ.output.0  2.27e-2     2.9ε
layers.22.self_attention.core_attention|g_activ.input.0   1.13e+0   145.8ε
layers.22.self_attention.core_attention|g_activ.input.1   6.94e-1    89.0ε
layers.22.self_attention.core_attention|g_activ.input.2   7.09e-1    91.0ε
layers.22.self_attention.linear_qkv|g_activ.output.0      7.19e-1    92.3ε
layers.22.input_layernorm|g_activ.input.0                 7.29e-1    93.5ε
...
layers.1|g_activ.output.0                                 9.34e-1   119.8ε
layers.1.mlp.linear_fc2|g_activ.output.0                  9.34e-1   119.8ε
layers.1.mlp.linear_fc2|g_activ.input.0                   9.36e-1   120.1ε
layers.1.mlp.linear_fc1|g_activ.output.0                  9.48e-1   121.6ε
layers.1.self_attention.linear_proj|g_activ.output.0      8.85e-1   113.5ε
layers.1.self_attention.core_attention|g_activ.output.0   8.63e-1   110.7ε
layers.1.self_attention.core_attention|g_activ.input.0    1.81e+0   232.5ε
layers.1.self_attention.core_attention|g_activ.input.1    1.99e+0   256.3ε
layers.1.self_attention.core_attention|g_activ.input.2    2.45e+0   314.7ε
layers.1.self_attention.linear_qkv|g_activ.output.0       2.44e+0   313.5ε
layers.1.input_layernorm|g_activ.input.0                  2.02e+0   259.2ε
layers.0|g_activ.output.0                                 1.37e+0   176.8ε
layers.0.mlp.linear_fc2|g_activ.output.0                  1.37e+0   176.8ε
layers.0.mlp.linear_fc2|g_activ.input.0                   1.37e+0   176.1ε
layers.0.mlp.linear_fc1|g_activ.output.0                  1.35e+0   174.2ε
layers.0.self_attention.linear_proj|g_activ.output.0      6.23e-1    80.0ε
layers.0.self_attention.core_attention|g_activ.output.0   6.35e-1    81.5ε
layers.0.self_attention.core_attention|g_activ.input.0    2.12e+0   272.1ε
layers.0.self_attention.core_attention|g_activ.input.1    2.30e+0   295.6ε
layers.0.self_attention.core_attention|g_activ.input.2    2.18e+0   280.6ε
layers.0.self_attention.linear_qkv|g_activ.output.0       2.18e+0   280.6ε
layers.0.input_layernorm|g_activ.input.0                  2.21e+0   283.4ε
```

- parameter gradient 

```
name                                                   relative_error  
decoder.layers.23.mlp.linear_fc2.bias                 4.20e-3     0.5ε
decoder.layers.23.mlp.linear_fc2.weight               5.16e-3     0.7ε
decoder.layers.23.mlp.linear_fc1.bias                 4.54e-3     0.6ε
decoder.layers.23.mlp.linear_fc1.weight               5.39e-3     0.7ε
decoder.layers.23.self_attention.linear_proj.bias     4.22e-3     0.5ε
decoder.layers.23.self_attention.linear_proj.weight   4.78e-3     0.6ε
decoder.layers.23.self_attention.linear_qkv.bias      6.67e-2     8.6ε
decoder.layers.23.self_attention.linear_qkv.weight    1.31e-1    16.9ε
decoder.layers.22.mlp.linear_fc2.bias                 1.33e-2     1.7ε
decoder.layers.22.mlp.linear_fc2.weight               2.45e-2     3.1ε
decoder.layers.22.mlp.linear_fc1.bias                 1.79e-2     2.3ε
decoder.layers.22.mlp.linear_fc1.weight               2.80e-2     3.6ε
decoder.layers.22.self_attention.linear_proj.bias     1.34e-2     1.7ε
decoder.layers.22.self_attention.linear_proj.weight   2.67e-2     3.4ε
decoder.layers.22.self_attention.linear_qkv.bias      8.05e-2    10.3ε
decoder.layers.22.self_attention.linear_qkv.weight    1.41e-1    18.1ε
...
decoder.layers.1.mlp.linear_fc2.bias                  4.83e-1    61.9ε
decoder.layers.1.mlp.linear_fc2.weight                4.69e-1    60.2ε
decoder.layers.1.mlp.linear_fc1.bias                  4.79e-1    61.4ε
decoder.layers.1.mlp.linear_fc1.weight                4.56e-1    58.6ε
decoder.layers.1.self_attention.linear_proj.bias      4.64e-1    59.6ε
decoder.layers.1.self_attention.linear_proj.weight    4.75e-1    61.0ε
decoder.layers.1.self_attention.linear_qkv.bias       7.98e-1   102.3ε
decoder.layers.1.self_attention.linear_qkv.weight     7.77e-1    99.7ε
decoder.layers.0.mlp.linear_fc2.bias                  6.11e-1    78.4ε
decoder.layers.0.mlp.linear_fc2.weight                6.19e-1    79.5ε
decoder.layers.0.mlp.linear_fc1.bias                  5.92e-1    75.9ε
decoder.layers.0.mlp.linear_fc1.weight                6.17e-1    79.1ε
decoder.layers.0.self_attention.linear_proj.bias      3.60e-1    46.2ε
decoder.layers.0.self_attention.linear_proj.weight    8.32e-1   106.7ε
decoder.layers.0.self_attention.linear_qkv.bias       7.40e-1    94.9ε
decoder.layers.0.self_attention.linear_qkv.weight     1.54e+0   198.3ε
embedding.position_embeddings.weight                  2.27e+0   291.4ε
embedding.word_embeddings.weight                      2.90e-1    37.3ε
```

- parameter `main_grad` before optimizer.step()

```
name                                                          relative_error 
embedding.word_embeddings.weight                              2.05e-1   26.4ε
embedding.position_embeddings.weight                          2.31e+0  296.3ε
decoder.layers.0.self_attention.linear_proj.weight            8.04e-1  103.2ε
decoder.layers.0.self_attention.linear_proj.bias              3.66e-1   47.0ε
decoder.layers.0.self_attention.linear_qkv.layer_norm_weight  1.41e+0  181.4ε
decoder.layers.0.self_attention.linear_qkv.layer_norm_bias    7.59e-1   97.4ε
decoder.layers.0.self_attention.linear_qkv.weight             1.50e+0  193.3ε
decoder.layers.0.self_attention.linear_qkv.bias               7.54e-1   96.7ε
decoder.layers.0.mlp.linear_fc1.layer_norm_weight             6.34e-1   81.4ε
decoder.layers.0.mlp.linear_fc1.layer_norm_bias               6.11e-1   78.4ε
decoder.layers.0.mlp.linear_fc1.weight                        5.94e-1   76.2ε
decoder.layers.0.mlp.linear_fc1.bias                          6.01e-1   77.1ε
decoder.layers.0.mlp.linear_fc2.weight                        6.10e-1   78.3ε
decoder.layers.0.mlp.linear_fc2.bias                          6.24e-1   80.0ε
decoder.layers.1.self_attention.linear_proj.weight            4.49e-1   57.7ε
decoder.layers.1.self_attention.linear_proj.bias              4.65e-1   59.7ε
decoder.layers.1.self_attention.linear_qkv.layer_norm_weight  8.50e-1  109.0ε
decoder.layers.1.self_attention.linear_qkv.layer_norm_bias    8.30e-1  106.5ε
decoder.layers.1.self_attention.linear_qkv.weight             7.79e-1  100.0ε
decoder.layers.1.self_attention.linear_qkv.bias               8.29e-1  106.4ε
decoder.layers.1.mlp.linear_fc1.layer_norm_weight             4.18e-1   53.6ε
decoder.layers.1.mlp.linear_fc1.layer_norm_bias               4.72e-1   60.6ε
decoder.layers.1.mlp.linear_fc1.weight                        4.26e-1   54.6ε
decoder.layers.1.mlp.linear_fc1.bias                          4.72e-1   60.6ε
decoder.layers.1.mlp.linear_fc2.weight                        4.39e-1   56.3ε
decoder.layers.1.mlp.linear_fc2.bias                          4.75e-1   60.9ε
...
decoder.layers.22.self_attention.linear_proj.weight           1.65e-2    2.1ε
decoder.layers.22.self_attention.linear_proj.bias             7.93e-3    1.0ε
decoder.layers.22.self_attention.linear_qkv.layer_norm_weight 9.53e-2   12.2ε
decoder.layers.22.self_attention.linear_qkv.layer_norm_bias   5.87e-2    7.5ε
decoder.layers.22.self_attention.linear_qkv.weight            9.53e-2   12.2ε
decoder.layers.22.self_attention.linear_qkv.bias              5.73e-2    7.4ε
decoder.layers.22.mlp.linear_fc1.layer_norm_weight            1.72e-2    2.2ε
decoder.layers.22.mlp.linear_fc1.layer_norm_bias              1.03e-2    1.3ε
decoder.layers.22.mlp.linear_fc1.weight                       1.75e-2    2.2ε
decoder.layers.22.mlp.linear_fc1.bias                         1.06e-2    1.4ε
decoder.layers.22.mlp.linear_fc2.weight                       1.52e-2    2.0ε
decoder.layers.22.mlp.linear_fc2.bias                         7.82e-3    1.0ε
decoder.layers.23.self_attention.linear_proj.weight           3.71e-3    0.5ε
decoder.layers.23.self_attention.linear_proj.bias             3.50e-3    0.4ε
decoder.layers.23.self_attention.linear_qkv.layer_norm_weight 7.68e-2    9.9ε
decoder.layers.23.self_attention.linear_qkv.layer_norm_bias   3.63e-2    4.7ε
decoder.layers.23.self_attention.linear_qkv.weight            8.50e-2   10.9ε
decoder.layers.23.self_attention.linear_qkv.bias              3.71e-2    4.8ε
decoder.layers.23.mlp.linear_fc1.layer_norm_weight            3.89e-3    0.5ε
decoder.layers.23.mlp.linear_fc1.layer_norm_bias              3.53e-3    0.5ε
decoder.layers.23.mlp.linear_fc1.weight                       3.89e-3    0.5ε
decoder.layers.23.mlp.linear_fc1.bias                         3.51e-3    0.5ε
decoder.layers.23.mlp.linear_fc2.weight                       3.83e-3    0.5ε
decoder.layers.23.mlp.linear_fc2.bias                         3.39e-3    0.4ε
decoder.final_layernorm.weight                                4.29e-3    0.6ε
decoder.final_layernorm.bias                                  3.46e-3    0.4ε
```

**Environment (please complete the following information):**
 - Megatron-LM commit ID: [`fdfcef87`](https://github.com/NVIDIA/Megatron-LM/commit/fdfcef87004205eca69ea462b375578185ea01f6)
 - PyTorch version: 2.5.1+cu121
 - CUDA version: 12.1
 - NCCL version: 2.21.5
 - TransformerEngine: [`754d2a0`](https://github.com/NVIDIA/TransformerEngine/commit/754d2a0023131f9e82309f87fe9fc182d1835da8) (current newest commit in `stable` branch)

**Proposed fix**
Check the `AttnFuncWithCPAndKVP2P.backward()` and fix the bug in it.

**Additional context**
Please fix it soon.  This is very serious bug.
